types/block_spec.md

```diff
@@ -0,0 +1,124 @@
+# Block and Header Validity
+
+## Abstract
+
+Like all blockchains, rollups are defined as the chain of **valid** blocks from the genesis, to the head. Thus, the block and header validity rules define the chain.
+
+Verifying a block / header is done in 3 parts:
+
+1. Verify correct serialization according to the protobuf spec
+
+2. Perform basic validation of the types
+
+3. Perform verification of the new block against the previous accepted block
+
+## Basic Validation
+
+Each type contains a `.ValidateBasic()` method, which verifies that certain basic invariants hold. The `ValidateBasic()` calls are nested, starting from the `Block` struct, all the way down to each subfield.
+
+The nested basic validation, and validation checks, are called as follows:
+
+```go
+Block.ValidateBasic()
+  // Make sure the block's SignedHeader passes basic validation
+  SignedHeader.ValidateBasic()
+    // Make sure the SignedHeader's Header passes basic validation
+    Header.ValidateBasic()
+	  verify ProposerAddress not nil
+	// Make sure the SignedHeader's Commit passes basic validation
+	Commit.ValidateBasic()
+	  // Ensure that someone signed the block
+	  verify len(c.Signatures) not 0
+	If sh.Validators is nil, or len(sh.Validators.Validators) is 0, assume based rollup, pass validation, and skip all remaining checks.
+	Validators.ValidateBasic()
+	  // github.com/rollkit/cometbft/blob/main/types/validator.go#L37
+	  verify sh.Validators is not nil, and len(sh.Validators.Validators) != 0
+	  // apply basic validation to all Validators
+	  for each validator:
+	    validator.ValidateBasic()
+		  validate not nil
+		  validator.PubKey not nil
+		  validator.VotingPower >= 0
+		  validator.Address == correct size
+	  // apply ValidateBasic to the proposer field:
+	  sh.Validators.Proposer.ValidateBasic()
+		validate not nil
+		validator.PubKey not nil
+		validator.VotingPower >= 0
+		validator.Address == correct size
+    Assert that SignedHeader.Validators.Hash() == SignedHeader.AggregatorsHash
+    Assert that len(SignedHeader.Commit.Signatures) == 1 (Exactly one signer check)
+	Verify the 1 signature
+  Data.ValidateBasic() // always passes
```

Comment:
Whats the purpose of this check if we don't do anything with it? Should we add something to Data.ValidateBasic() ?

Reply:
Issue created
https://github.com/rollkit/rollkit/issues/1271